src/stream/src/executor/backfill/snapshot_backfill.rs

```diff
@@ -0,0 +1,494 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+use std::future::{pending, Future};
+use std::mem::replace;
+
+use anyhow::anyhow;
+use futures::future::Either;
+use futures::{pin_mut, Stream, TryStreamExt};
+use itertools::Itertools;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::catalog::TableId;
+use risingwave_common::row::OwnedRow;
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+use risingwave_hummock_sdk::HummockReadEpoch;
+use risingwave_storage::error::StorageResult;
+use risingwave_storage::table::batch_table::storage_table::StorageTable;
+use risingwave_storage::table::ChangeLogRow;
+use risingwave_storage::StateStore;
+use tokio::select;
+use tokio::sync::mpsc::UnboundedReceiver;
+
+use crate::executor::backfill::utils::{create_builder, mapping_chunk};
+use crate::executor::prelude::{try_stream, StreamExt};
+use crate::executor::{
+    expect_first_barrier, ActorContextRef, BackfillExecutor, Barrier, BoxedMessageStream, Execute,
+    Executor, Message, Mutation, StreamExecutorError, StreamExecutorResult,
+};
+use crate::task::CreateMviewProgress;
+
+pub struct SnapshotBackfillExecutor<S: StateStore> {
+    /// Upstream table
+    upstream_table: StorageTable<S>,
+
+    /// Upstream with the same schema with the upstream table.
+    upstream: Executor,
+
+    /// The column indices need to be forwarded to the downstream from the upstream and table scan.
+    output_indices: Vec<usize>,
+
+    progress: CreateMviewProgress,
+
+    chunk_size: usize,
+    rate_limit: Option<usize>,
+
+    barrier_rx: UnboundedReceiver<Barrier>,
+}
+
+impl<S: StateStore> SnapshotBackfillExecutor<S> {
+    #[expect(clippy::too_many_arguments)]
+    pub fn new(
+        upstream_table: StorageTable<S>,
+        upstream: Executor,
+        output_indices: Vec<usize>,
+        _actor_ctx: ActorContextRef,
+        progress: CreateMviewProgress,
+        chunk_size: usize,
+        rate_limit: Option<usize>,
+        barrier_rx: UnboundedReceiver<Barrier>,
+    ) -> Self {
+        if let Some(rate_limit) = rate_limit {
+            debug!(
+                rate_limit,
+                "create snapshot backfill executor with rate limit"
+            );
+        }
+        Self {
+            upstream_table,
+            upstream,
+            output_indices,
+            progress,
+            chunk_size,
+            rate_limit,
+            barrier_rx,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(mut self) {
+        debug!("snapshot backfill executor start");
+        let mut upstream = self.upstream.execute();
+        let upstream_table_id = self.upstream_table.table_id();
+        let first_barrier = expect_first_barrier(&mut upstream).await?;
```

Comment:
Do we still need this? I would think we just use the first_recv_barrier below.

Reply:
In the current implementation, we still receive all the upstream messages. We simply discard all the stream chunks when consuming snapshot and log store. We should still receive and buffer the upstream barriers. First, we need to compare the first barrier received from upstream and from local barrier worker to see whether the executor is recovered or newly created and decide whether to backfill. Second, we need to compare the epoch in barriers received from both side to ensure equality.